### ISSUE TITLE
bare-except (W0702) tricks developers into writing bugs

### ISSUE THREAD

**Issue by couling** [PyCQA/pylint#4807]
### Current problem

_Apologies if I'm quite forthright with this one.  This particular problem has burned me pretty hard and I now suspect it has my colleagues also.  I'm aware of #3656 but that doesn't really address the problem._

Pylint is tricking developers into writing `except Exception:` where they really did mean (and most likely did want)`except:`.  The explanation of "bare-except" W0702 is:

> Catching exceptions should be as precise as possible. The type of exceptions that can be raised should be known in advance. Using a catch-all Exception instance defeats the purpose of knowing the type of error that occurred, and prohibits the use of tailored responses.

But, specifically where code was originally `except:` ... `raise` (bare-raise), there are too many cases where this description is entirely wrong.  

Indeed in my own code it's incorrect in almost every case.  That's because `except:`...`raise` is responding to the fact `try:` block did not complete and it isn't "handle" the exception at all.  Where the bare-raise is the only way out of the except block, there's almost never any kind of tailored response wanted.

The current set of pylint linting errors is tricking developers into refactoring `except:` ... `raise` into `except Exception:` ... `raise`.  **This does NOT do what the developer intended**.  If the developer's intent was to do something if-and-only-if the `try:` block did not complete, then `except Exception:` will miss some common events such as the user hitting ctrl^c at the command line.

## Consider these two high-level examples
```python
try:
    do_something(x)
except:
    # Where cleanup(x) cannot execute on success, this must not go into finally
    cleanup(x)
    raise
```
We definitely do want to `cleanup()` irrespective of why `do_something()` failed.  

```python
try:
    do_something(x)
except:
    # Despite the fact we are logging to explain why do_something failed...
    # The condition that we want to log the error is dependent on there being ANY error.
    logger.error("Failed to complete do_something", ex_info=True)
    raise
```
We definitely do want to log _something_ irrespective of why `do_something()` failed.  Yes the log is reporting the error to the log so in that sense it is responding to the specific error.  But the actual code path in every case is the same.

### Desired solution

The simplest thing would be to relax W0702 so that it does not trigger if the exception is immediately re-raised.  I believe a very similar clause applies to warnings associated with `except Exception:`.  Specifically it should not warn where all code paths lead to a bare `raise`.

That is
```python
try:
    something()
except:
    something_else()
```
would still raise the same warning it always has.  But 
```python
try:
    something()
except:
    something_else()
    raise
```
should no.

## Going further
_Poking my head above the parapet ... I might be shot at for this_.  I wonder if an inverse warning would be appropriate on this code.  Something along the lines of "cleanup code may not execute in all intended circumstances":

```python
try:
    something()
except Exception:
    something_else()
    raise
```

To be clear, that's based on the condition of an `except Exception:` block ending all code-paths with a bare `raise`.

I realise this would be opposed to many other linting tools as well as raising a lot of warnings on code that pylint has already pushed people into writing.  So I sense there may be some hot feelings about this suggestion.

### Additional context

I'll give a pratical example and discuss some connected issues.  Common enough is the task of writing a file atomically.  Since no operating systems support this.  Two approaches to do this:

1. Write the file in a temporary directory on the same file system and then move (rename) the file to its final resting place.
2. place a "lock" file in the directory preventing other threads reading while the file is incomplete.

In neither scenario would we want to abort before the operation has either completed or rolled back entirely.  In 1) it would accumulate unwanted files in the temp directory.  In 2) much more dangerously it would leave the lock file, which if deleted by the user would also leave a half written file in the directory, potentially causing corruption further down the line. 

So for 1) the following code might be okay(ish):
```python
def write_attomic(content: bytes, target_path: str, temp_directory: str):
    file = NamedTemporaryFile(mode='wb', dir=temp_directory, delete=False)
    try:
        with file:
            file.write(content)
        os.rename(file.file)
    except:
        os.unlink(file.file)
        raise
```
Of course it does have one fault where the unlink could fail.  Thus to do this "safely" we end up with the somewhat hidious:
```python
def write_attomic(content: bytes, target_path: str, temp_directory: str):
    file = NamedTemporaryFile(mode='wb', dir=temp_directory, delete=False)
    try:
        with file:
            file.write(content)
        os.rename(file.file)
    except:
        try:
            os.unlink(file.file)
        except Exception:  # or bare except? - original exception even if we failed cleanup
            pass
        raise
```

On balance I would be comfortable with either of these in code.

However:
```python
def write_attomic(content: bytes, target_path: str, temp_directory: str):
    file = NamedTemporaryFile(mode='wb', dir=temp_directory, delete=False)
    try:
        with file:
            file.write(content)
        os.rename(file.file)
    except Exception:
        os.unlink(file.file)
        raise
```
... If the user hits ctrl-c then we have a problem.

**Comment by couling** [comment]
After filing this I read what PEP 8 says.  Gratifyingly PEP 8 calls out the same two specific exceptions that I've described in my high-level examples: logging and cleanup

> A good rule of thumb is to limit use of bare 'except' clauses to two cases:
>
> - If the exception handler will be printing out or logging the traceback; at least the user will be aware that an error has occurred.
> - If the code needs to do some cleanup work, but then lets the exception propagate upwards with raise. try...finally can be a better way to handle this case.

I don't believe pylint can accurately identify what is [not] logging or cleanup code.  However I believe the detection of a bare `raise` should be a good enough rule of thumb.

Further I really think that false positives on the exceptions PEP 8 explicitly allows is more damaging than false negatives are helpful.

**Comment by Pierre-Sassoulas** [comment]
I agree with you mostly, this check is annoying and almost always a false positive for someone experienced, and almost always misunderstood for someone inexperienced. First step imo: making it optional in #3512. 

Second step should be discussed. I think there's two kind of possibility to have this kind of code:
- Inexperienced dev don't want errors to be raised, so naively suppress them and silently suppressing exception makes the code 3 magnitude harder to troubleshoot. This is probably the thing the checker was aiming at.
- Experienced dev catch everything in a routine so the routine is not failing when a single file other a thousands fail. This is where the false positive come from.

I think you described in detail why the current message is misleading and promote the wrong solution so I won't paraphrase it. 

Now about the solution. Maybe once it's optional only the most experienced user will use it and then the problem solve itself because they'll know if they need to disable it or not. So doing nothing more is an option.

I think but I might be wrong that suggesting to use bare except would introduce new false positive and probably new confusion. In particular I don't think everyone is aware that bare except catch even KeyboardInterupt and I don't think that preventing ctrl+C is the expected behavior in most case. Python is used for script and not being able to interrupt a script is very annoying.. 

What about merging `except:` and `except Exception:` and making the message more detailed to help inexperienced dev using code base with the check enabled understands the issue ?

**Comment by couling** [comment]
I've struck through "designed solution 1" it was written in error - apologies.

Actually I'll hold my hands up and admit I got a little confused about which linter had raised issues with precisely which variant of my code before things stopped complaining at me.  Yet boiling it down pylint inadvertently perpetuates a particular myth: the idea that adding an explicit exception fixes the core problem.

So @Pierre-Sassoulas  Yes I think merging the two would really help prevent pylint perpetuating that myth.

Alternatively, specifically warning the user about the ctrl^c issue and mentioning that it's acceptable where the exception is re-raised would help also.

My concern that developers are tricked into switching  from `except:` to `except: Exception`.  Other linters<sup>1</sup> have different behaviour around `except:` vs `except Exception:` where the warning is issued despite a `raise` for `except:`.  I'm leaving in my "Desired Solution 2" because it helps developers understand that other linter's interpretation of this is plainly wrong and will break code.

**Comment by Pierre-Sassoulas** [comment]
Blocked by #3512
